lib/connextclient/ConnextClient.ts

```diff
@@ -0,0 +1,699 @@
+  ExpectedIncomingTransfer,
+  ProvidePreimageEvent,
+  TransferReceivedEvent,
+} from './types';
+import { parseResponseBody } from '../utils/utils';
+import { Observable, fromEvent } from 'rxjs';
+import { take, pluck, timeout, filter } from 'rxjs/operators';
+
+export const TIMELOCK_BUFFER = 100;
+
+const MAX_AMOUNT = Number.MAX_SAFE_INTEGER;
+
+interface ConnextClient {
+  on(event: 'preimage', listener: (preimageRequest: ProvidePreimageEvent) => void): void;
+  on(event: 'transferReceived', listener: (transferReceivedRequest: TransferReceivedEvent) => void): void;
+  on(event: 'htlcAccepted', listener: (rHash: string, amount: number, currency: string) => void): this;
+  on(event: 'connectionVerified', listener: (swapClientInfo: SwapClientInfo) => void): this;
+  once(event: 'initialized', listener: () => void): this;
+  emit(event: 'htlcAccepted', rHash: string, amount: number, currency: string): boolean;
+  emit(event: 'connectionVerified', swapClientInfo: SwapClientInfo): boolean;
+  emit(event: 'initialized'): boolean;
+  emit(event: 'preimage', preimageRequest: ProvidePreimageEvent): void;
+  emit(event: 'transferReceived', transferReceivedRequest: TransferReceivedEvent): void;
+}
+
+/**
+ * Waits for the preimage event from SwapClient for a specified hash
+ * @param client the swap client instance to listen events from
+ * @param expectedHash the expected hash of the payment
+ * @param errorTimeout optional maximum duration of time to wait for the preimage
+ */
+const waitForPreimageByHash = (
+  client: SwapClient,
+  expectedHash: string,
+  errorTimeout = 30000,
+): Promise<string> => {
+  // create an observable that emits values when a preimage
+  // event is triggered on the client
+  const preimage$: Observable<ProvidePreimageEvent> = fromEvent(client, 'preimage');
+  const expectedPreimageReceived$ = preimage$.pipe(
+    // filter out events that do not match our expected hash
+    filter(preimageEvent => preimageEvent.rHash === expectedHash),
+    // map the ProvidePreimageEvent object to a string as the consumer is
+    // only interested in the value of the preimage
+    pluck('preimage'),
+    // complete the observable and clean up all the listeners
+    // once we receive 1 event that matches our conditions
+    take(1),
+    // emit an error if the observable does not emit any values
+    // for the specified duration
+    timeout(errorTimeout),
+  );
+  // convert the observable to a promise that resolves on complete
+  // and rejects on error
+  return expectedPreimageReceived$.toPromise();
+};
+
+/**
+ * A class representing a client to interact with connext.
+ */
+class ConnextClient extends SwapClient {
+  public readonly type = SwapClientType.Connext;
+  public readonly finalLock = 200;
+  public address?: string;
+  /** A map of currency symbols to token addresses. */
+  public tokenAddresses = new Map<string, string>();
+  private expectedIncomingTransfers = new Map<string, ExpectedIncomingTransfer>();
```

Comment:
It'd be good to have a comment here I think to explain that this is basically the equivalent of invoices.